### PR TITLE
resolver: Make `ResolverFuture::lookup_ip` lazier

### LINKED ISSUE #430: `ResolverFuture::lookup_ip()` is not lazy (enough)
Futures shouldn't do any work until polled, so that they can support code patterns like this (taken from real code):

```rust
        // Construct the future, expecting nothing will happen yet.
        let lookup = self.resolver.lookup_ip(&name.0);

        // Do something that needs to be done before the lookup
        // is initiated.
        let f = Timeout::new(delay, &self.executor)
            .expect("Timeout::new() won't fail")
            .then(move |_| {
                // OK, now actually start doing the lookup work.
                lookup
            });

        // Wrap the future in a Box, giving it a lifetime that is (potentially)
        // longer than `self`. Because of this lifetime issue, we can't call
        // `self.resolver.lookup_ip(&name.0);` inside the closure above.
        return Box::new(f);
```

`ResolverFuture::lookup_ip()` appears to be doing a lot of work and so it doesn't seem to support this code pattern well.

### PR DESCRIPTION
**Pull request by hawkw**
This PR refactors the implementation of `ResolverFuture::lookup_ip` to lazily
prepare the `LookupIpFuture` when the returned future starts executing, rather
than eagerly when that function is called.

I took the simpler approach here and simply wrapped the preparation in a
`future::lazy`, rather than changing the implementation of `LookupIpFuture`.
This is because a number of Trust-DNS' tests construct `LookupIpFuture`s
without first constructing a `ResolverFuture`, and changing `LookupIpFuture`
to do all the preparation done by `ResolverFuture::lookup_ip` would introduce
a significant amount of additional complexity to avoid breaking these tests.

Notably, the use of `future::lazy` here changes the return type from
`ResolverFuture::lookup_ip` to `impl Future<...>` rather than a concrete type.
If this API change is problematic downstream, we could add a new future type
that does the preparation and then becomes a `LookupIpFuture`, like I did
originally in hawkw/trust-dns@a76769e25f075414d6aeae246168afc539506cbd. I found
that this change makes the _code_ more complex, but if being able to return a
concrete named type here presents a singificantly better _API surface_, I'd be
happy to change this branch to do that instead.

Closes #430

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

/cc @briansmith

### CODE REVIEW

**Comment by briansmith** [comment]
> This is because a number of Trust-DNS' tests construct LookupIpFutures
> without first constructing a ResolverFuture

This sounds bad. IMO the tests should be using the API the way the users of the library would use it, exactly so they don't depend on the kind of details that make this refactoring difficult. That said, I don't know how appropriate it would be to change those tests to construct a `ResolverFuture` first.

This might be a good intermediate step but it's not a good end-state.

**Comment by hawkw** [comment]
@briansmith
There are a number of tests in `resolver/src/resolver_future.rs` that do exercise the `ResolverFuture::lookup_ip()` API, so it's not that that API is untested: https://github.com/bluejekyll/trust-dns/blob/fb56517caa05569c72c0a16586703d73fdfc1517/resolver/src/resolver_future.rs#L434-L888

However, there are _also_ several tests in `integration-tests/tests/lookup_tests.rs` that construct `LookupIpFuture`s with `LookupIpFuture::lookup()`, rather than using `ResolverFuture`: 
https://github.com/bluejekyll/trust-dns/blob/fb56517caa05569c72c0a16586703d73fdfc1517/integration-tests/tests/lookup_tests.rs#L62-L239

I wasn't wholly aware of the intention behind these tests, so I didn't think I should modify them.

**Comment by bluejekyll** [comment]
Not sure if his helps, but the intention behind those tests is to isolate specific details about the LookupIpFuture from the Resolver logic (like NameServerPool), unit vs integration tests. The reason they are in the integration library is to reuse the TestClient and MockClient.

**Comment by briansmith** [comment]
OK. I don't have a strong opinion either way, as long as `lookup_id()` is lazy. I think it's really up to @bluejekyll's preferences as the maintainer.

**Comment by bluejekyll** [comment]
I’m at a conference at the moment, I’ll try and find some time tonight to look at this in detail. Apologies on the slow response. This will be true over the next couple of days.

**Comment by hawkw** [comment]
@bluejekyll That's fine, I'm happy to wait until you have the time to review this branch in depth. :)

**Comment by bluejekyll** [comment]
Yes. I’ve heen trying to stabilize some of these tests. It looks like there are still some gaps.

File: resolver/examples/global_resolver.rs [hunk 1]
```diff
@@ -88,7 +88,11 @@ lazy_static! {
 ///
 /// This looks up the `host` (a &str or String is good), and combines that with the provided port
 ///   this mimics the lookup functions of std::net.
-pub fn resolve<N: IntoName + TryParseIp>(host: N, port: u16) -> IoFuture<Vec<SocketAddr>> {
+pub fn resolve<N>(host: N, port: u16) -> IoFuture<Vec<SocketAddr>>
+where
+    // TODO: It would be nice to remove the `'static` bound on `N`.
+    N: IntoName + TryParseIp + Send + 'static,
```
Comment (bluejekyll) [inline resolver/examples/global_resolver.rs]: I’m not clear why the 'static bound is needed here.

We could do the name conversion before passing into the lazy if that’s the issue, right?

**Pull request closed without merging**